src/mempool/storage.rs

```diff
@@ -405,9 +351,52 @@ impl Storage {
                             return DataProposalVerdict::Refuse;
                         }
                     };
+                    // TODO: figure out how to generalize this
+                    let is_recursive = proof_tx.contract_name.0 == "risc0-recursion";
+
+                    if is_recursive {
+                        let Some(recursive_metadata) = &proof_tx.recursive_metadata else {
+                            warn!("Refusing DataProposal: recursive proof transaction missing metadata");
+                            return DataProposalVerdict::Refuse;
+                        };
+                        match verify_recursive_proof(&proof_bytes, verifier, program_id) {
+                            Ok((program_ids, hyle_outputs)) => {
+                                let program_ids_match =
+                                    std::iter::zip(program_ids.iter(), recursive_metadata.iter())
+                                        .all(|(output, metadata)| output == metadata);
+                                let outputs_match = std::iter::zip(
+                                    hyle_outputs.iter(),
+                                    proof_tx.proven_blobs.iter(),
+                                )
+                                .all(
+                                    |(output, BlobProofOutput { hyle_output, .. })| {
+                                        output == hyle_output
+                                    },
+                                );
+                                if hyle_outputs.len() != proof_tx.proven_blobs.len()
+                                    || !outputs_match
+                                    || !program_ids_match
+                                {
+                                    warn!("Refusing DataProposal: incorrect HyleOutput in proof transaction");
+                                    return DataProposalVerdict::Refuse;
+                                }
+                            }
+                            Err(e) => {
+                                warn!("Refusing DataProposal: invalid proof transaction: {}", e);
+                                return DataProposalVerdict::Refuse;
+                            }
+                        }
+                    }
                     match verify_proof(&proof_bytes, verifier, program_id) {
-                        Ok(_) => {
-                            if proof.hash() != proof_tx.proof_hash {
+                        Ok(outputs) => {
+                            // TODO: we could check the blob hash here too.
+                            if outputs.len() != proof_tx.proven_blobs.len()
+                                && std::iter::zip(outputs.iter(), proof_tx.proven_blobs.iter()).any(
```

Comment:
```suggestion
                                || std::iter::zip(outputs.iter(), proof_tx.proven_blobs.iter()).any(
```
ou j'ai mal compris?

Reply:
Bien vu